confluence-xml/src/main/java/org/xwiki/contrib/confluence/filter/internal/macros/AbstractMacroConverter.java

```diff
@@ -54,13 +63,80 @@ public void toXWiki(String confluenceId, Map<String, String> confluenceParameter
 
         printUnhandledInfo(confluenceId, tracedParameters);
 
         listener.onMacro(id, parameters, content, inline);
     }
 
+    private Map<String, String> maybeKeepConfluenceParameters(Map<String, String> confluenceParameters,
+        Map<String, String> parameters, Collection<String> parametersWithUnhandledValues,
+        Collection<String> unhandledParameters)
+    {
+        if (inputContext == null) {
+            logger.info("Could not determine the keeping parameter mode, assuming NONE");
+            return parameters;
+        }
+
+        ConfluenceInputProperties properties = inputContext.getProperties();
+        String mode = properties.getKeptMacroParameterMode();
+        if (StringUtils.isEmpty(mode) || mode.equals("NONE")) {
+            return parameters;
+        }
+
+        String prefix = properties.getKeptMacroParameterPrefix();
+
+        Map<String, String> newParameters = null;
+        if (mode.equals("ALL")) {
+            for (Map.Entry<String, String> confluenceParameter : confluenceParameters.entrySet()) {
+                newParameters = maybeAddKeptParameter(parameters, prefix, newParameters,
+                    confluenceParameter.getKey(), confluenceParameter.getValue());
+            }
+        } else if (mode.equals("UNHANDLED")) {
+            newParameters = addUnhandledParameters(confluenceParameters, parameters, parametersWithUnhandledValues,
+                newParameters, prefix);
+            newParameters = addUnhandledParameters(confluenceParameters, parameters, unhandledParameters,
+                newParameters, prefix);
+        } else {
+            logger.error("Unexpected keeping parameter mode [{}]. Assuming NONE. This should not happen.", mode);
+        }
+
+        return newParameters == null ? parameters : newParameters;
+    }
+
+    private static Map<String, String> addUnhandledParameters(Map<String, String> confluenceParameters,
+        Map<String, String> parameters, Collection<String> parametersWithUnhandledValues,
+        Map<String, String> newParameters, String prefix)
+    {
+        Map<String, String> np = newParameters;
+        for (String parameterName : parametersWithUnhandledValues) {
+            String parameterValue = confluenceParameters.get(parameterName);
+            np = maybeAddKeptParameter(parameters, prefix, newParameters, parameterName, parameterValue);
+        }
+        return np;
+    }
+
+    private static Map<String, String> maybeAddKeptParameter(Map<String, String> parameters,
+        String prefix, Map<String, String> newParameters, String parameterName, String parameterValue)
+    {
+        Map<String, String> np = newParameters;
+        String keptParameterName =
+            (ATLASSIAN_MACRO_OUTPUT_TYPE.equals(parameterName) || parameterName.startsWith(prefix))
+            ? parameterName
+            : prefix + parameterName;
+        if (!parameters.containsKey(parameterName) && !parameters.containsKey(keptParameterName)) {
```

Comment:
So If I understand correctly, we avoid to duplicate the parameter ? By example if I've in confluence `a="xx"` and the converter return a parameter `a="yy"`, we won't add a parameter `a="xx"` even if we have `keptMacroParameterMode=ALL` ?

Also I was thinking that maybe when we have `keptMacroParameterMode=ALL` it would be useful to log a warning if we don't save the parameter, in case we have a conflict. In most of case it should not happen, but in case it happen, It could be useful to know that something strange this happen.